Library/OcCpuLib/FrequencyDetect.c

```diff
@@ -225,6 +228,57 @@ InternalCalculateTSCFromPMTimer (
   return TSCFrequency;
 }
 
+UINT64
+InternalCalculateTSCFromApplePlatformInfo (
+  OUT  UINT64  *FSBFrequency  OPTIONAL
+  )
+{
+  EFI_STATUS                             Status;
+  APPLE_PLATFORM_INFO_DATABASE_PROTOCOL  *PlatformInfo;
+  UINT32                                 Size;
+  UINT64                                 FsbFreq;
+
+  Status = gBS->LocateProtocol (
+    &gApplePlatformInfoDatabaseProtocolGuid,
+    NULL,
+    (VOID **) &PlatformInfo
+    );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCCPU: Failed to locate ApplePlatformInfo protocol - %r\n", Status));
+    return 0;
+  }
+
+  Status = PlatformInfo->GetFirstDataSize (
+    PlatformInfo,
+    &gAppleFsbFrequencyPlatformInfoGuid,
+    &Size
+    );
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "OCCPU: Failed to get first data size - %r\n", Status));
+    return 0;
+  }
+
+  Status = PlatformInfo->GetFirstData (
```

Comment:
You should check that `Size` is not bigger than `sizeof (UINT64)` before calling `GetFirstData` here. Otherwise it will overwrite wrong memory. For our values it is an error, so it is safe to just return 0 if that actually happens.